### DISCUSSION TITLE
Possibility to trigger a KNX scene based on a HomeKit scene and vice versa?

### DISCUSSION THREAD

**Discussion by pfeifix** [brutella/hkknx-public#312] [Ideas]
Für die einheitliche Szenengestaltung zwischen KNX und HomeKit wäre es super, wenn es eine Möglichkeit für folgendes gäbe:

- HomeKit-Szene wird ausgelöst -> KNX-Szene wird getriggert
- KNX-Szene wird ausgelöst ->HomeKit-Szene wird getriggert

Also ein Mapping zwischen KNX- und Homekit-Szenen.

Gesetzt den Fall, dass das HomeKit-API entsprechende Events anbietet, um Szenenumschaltungen mitzubekommen, könnte man das Setzen der KNX-Szene bspw. durch eine Wartebedingung in einem hkknx-Skript ermöglichen.

Wenn beides funktioniert (Mapping hin und zurück), könnte man mit diesem Mechanismus super elegant Anzeigen von physischen KNX-Tastern oder Bedienzentralen konsistent halten. In meinem Fall verwalte ich Szenen komplett in HomeKit. Wenn meine KNX-Bedienelemente nun aber das umschalten dieser (HomeKit-)Szenen erlauben würden, wäre das für diesen Anwendungsfall ziemlich genial. 

(Wobei mir klar ist, dass das Szenenkonzept von KNX technisch anders arbeitet. Für mich ist es aber nicht zielführend, alle Zielzustände der HomeKit-Szenen noch einmal zu replizieren, das ist aufwändig und fehleranfällig, wenn es überhaupt abschließend möglich wäre. HomeKit-Trigger wären im KNX ja nie zugänglich.)

**Comment by brutella** [comment]
Eine HomeKit-Szene kann eine KNX-Szene auslösen, indem in hkknx ein Schalter mit einer Szenen-Gruppenadressen verknüpft wird. Die HomeKit-Szene schaltet dann diesen Szenen-Schalter ein und löst somit eine KNX-Szenen aus.

Andersrum geht das auch, indem in hkknx ein Taster mit einer Szenen-Gruppenadressen verknüpft wird. Ein Taster kann durch mehrere KNX-Szenen ausgelöst werden, da bis zu 8 Tasten mit jeweils 3 Funktionen (Einmal, Zweimal, Lang gedrückt) zur Verfügung stehe. Wird die Szene in KNX ausgelöst, wird die jeweilige Funktion (Einmal, Zweimal, Lang gedrückt) in HomeKit ausgelöst. In HomeKit kann man damit eine Szene auslösen, indem man den Taster als Auslöser in einer HomeKit Automation festlegt und als Aktion eine HomeKit Szene definiert.

Leider gibt es technisch zurzeit keine andere Lösung, da HomeKit-Szenen für hkknx nicht direkt zur Verfügung stehen. Es muss ein Umweg über einen Taster oder einen Schalter gemacht werden.

**Comment by pfeifix** [comment]
Ich habe das jetzt mal so eingebaut.

1. HomeKit  -> KNX:
Schalter mit Szenenobjekt angelegt. HomeKit-Szene so konfiguriert, dass sie diesen Schalter auslöst (einschaltet).

2. KNX -> HomeKit:
Taster mit gleichem Szenenobjekt angelegt. Im HomeKit am Taster die HomeKit-Szene konfiguriert.

Ergebnis: Wenn ich jetzt über die Home-App die Szene einschalte, wird sie kurz aktiviert, danach wieder deaktiviert. Auf jeden Fall wird sie nicht als aktiviert angezeigt. Das KNX-Szenentelegramm wird gesendet.

Kann es sein, dass HomeKit die Feedback-Schleife als Fehler erkennt und deshalb die Szene nicht angeschaltet lässt?

Könnte man das unterbinden, indem die Implementierung prüft, ob das Drücken des Tasters über einen Teilnehmer am Bus ausgelöst wird oder durch hkknx selbst (z. B. über die physikalische Quelladresse im KNX-Telegramm) und dann den HomeKit-Taster nur für Busteilnehmer betätigt?

**Reply by brutella** [reply]
> Ergebnis: Wenn ich jetzt über die Home-App die Szene einschalte, wird sie kurz aktiviert, danach wieder deaktiviert. Auf jeden Fall wird sie nicht als aktiviert angezeigt. Das KNX-Szenentelegramm wird gesendet.

Das Problem ist, dass in HomeKit Szenen aktiv sein können. Das ist der Fall, wenn der aktuelle Status der Geräte, die in der Szene definiert sind, mit dem Status in der Szene übereinstimmen. Also wenn eine Szene die Helligkeit einer Lampe auf 50 % stellt, dann wird in HomeKit die Szene als "aktiv" angezeigt, wenn die Lampe 50% hat.

In KNX wiederum gibt es das Konzept von einer aktiven Szene nicht. Eine Szene wird auf den KNX-Bus gesendet, kann aber nicht ausgelesen werden. Daher kann man bei KNX nicht herausfinden, welche Szene gerade aktiv ist.

Wenn du unbedingt in HomeKit sehen willst, ob eine Szene aktiv ist oder nicht, dann müsstest du in HomeKit die einzelnen Geräte und deren Zustände in einer Szene speichern. HomeKit senden beim aktivieren der Szene nicht die Szenennummer sondern die Zustände der Geräte.

**Reply by pfeifix** [reply]
Mir geht es gar nicht mal darum, dass ich den Szenenzustand (an/aus) in KNX abbilden möchte. Alles, was ich erwarten würde, wäre, wenn ich die KNX-Szene über das Einschalten des hkknx-Szenen-Schalter triggere, dass sie dann auch in HomeKit (über die Notification durch den Szenen-Taster) aktiviert wird (oder aktiv bleibt, wenn sie vorher schon an war). Natürlich kennt KNX keine "Aus-Szene". Aber die Taster nach dem Empfangen des Szenenobjekts sollte im HomeKit die Szene schon immer aktivieren und zwar unabhängig davon, ob das Schalten der HomeKit-Szene selbst vorher zum Aussenden des KNX-Szenen-Telegramms geführt hat. Bei mir wird sie nach dem oben beschriebenen Vorgehen kurz aktiviert und anschließend wieder ausgeschaltet - das leuchtet mir im Moment nicht ein, warum das so sein sollte.
Ich habe auch noch andere Geräte in der Szene, deren Zustände nach wie vor jeweils der eingeschalteten HomeKit-Szene entsprechen, also müsste die Szene als angeschaltet angezeigt werden.

**Reply by brutella** [reply]
> Bei mir wird sie nach dem oben beschriebenen Vorgehen kurz aktiviert und anschließend wieder ausgeschaltet - das leuchtet mir im Moment nicht ein, warum das so sein sollte.

Das ist deshalb so, weil der Schalter nach dem Einschalten von hkknx automatisch wieder ausgeschaltet wird. Wenn das nicht gemacht werden würde, hätte das folgende Auswirkungen.
- In HomeKit würde es so aussehen, als ob die Szene aktiv ist. Das stimmt aber nur, solange die Geräte im Szenenzustand bleiben. Weil aber HomeKit nicht weiss, welche Geräte mit der Szene angesprochen werden, kann es nicht wissen ob die Szene aktiv ist oder nicht. Somit zeigt HomeKit etwas falschen was – das kann die User verwirren.
- Du müsstest den Schalter wieder ausschalten, um die Szene erneut senden zu können.

Jetzt wo ich darüber so nachdenke, könnte man einen Zwischenweg gehen. Man könnte den Szenenschalter solange eingeschaltet lassen, solange keine neue Szenennummer an die angegebene Gruppenadresse gesendet wird. Somit bliebe der Szenenschalter nach dem einschalten aktiv. Wird eine andere Szenennummer (über KNX oder HomeKit) an die Gruppenadresse gesendet, schaltet sich der Schalter wieder aus. 🤔

**Reply by pfeifix** [reply]
Die Idee mit dem Zwischenweg hört sich super an. Wenn sich das umsetzen ließe, fände ich das sehr hilfreich. Während keine andere Szenennummer auf dem KNX kommuniziert wird, kann HomeKit ja davon ausgehen, dass die alte noch "aktiv" ist.

**Reply by mbrockeu** [reply]
Wie im „Bug“ beschrieben habt ihr gerade ne Funktion gekillt die damals eingebaut wurde. 🤷🏻‍♂️
